File: rust/a2021/src/bin/10.rs
fn main() {
    let mut part1 = 0;
    let mut part2: Vec<u64> = Vec::new();
    'll: for line in common::get_problem_lines(2021, 10) {
        let mut stack: Vec<u8> = Vec::new();
        for c in line.bytes() {
            match c {
                b'(' => stack.push(b'('),
                b'[' => stack.push(b'['),
                b'{' => stack.push(b'{'),
                b'<' => stack.push(b'<'),
                b')' => {
                    if stack.pop() != Some(b'(') {
                        part1 += 3;
                        continue 'll;
                    }
                }
                b']' => {
                    if stack.pop() != Some(b'[') {
                        part1 += 57;
                        continue 'll;
                    }
                }
                b'}' => {
                    if stack.pop() != Some(b'{') {
                        part1 += 1197;
                        continue 'll;
                    }
                }
                b'>' => {
                    if stack.pop() != Some(b'<') {
                        part1 += 25137;
                        continue 'll;
                    }
                }
                _ => panic!("other character"),
            }
        }
        let mut ans = 0;
        for c in stack.iter().rev() {
            ans = ans * 5
                + match c {
                    b'[' => 2,
                    b'(' => 1,
                    b'{' => 3,
                    b'<' => 4,
                    _ => panic!("A"),
                }
        }
        part2.push(ans);
    }
    println!("part 1: {}", part1);
    part2.sort();
    println!("part 2: {}", part2[part2.len() / 2]);
}

File: python/a2024/10.py
from problem import get_problem, get_problem_lines
from collections import defaultdict

m = defaultdict(lambda:100)
starts = []

for i, line in enumerate(get_problem_lines()):
    for j, c in enumerate(line):
        if c == '0':
            starts.append(i+1j*j)
        m[i+1j*j] = int(c)

ans = 0
ans2 = 0
for z in starts:
    reachable = set()
    stack = [z]
    while stack:
        loc = stack.pop()
        if m[loc] == 9:
            ans2 += 1
            reachable.add(loc)
        for loc_ in (loc+1, loc-1, loc+1j, loc-1j):
            if m[loc_] - m[loc] == 1:
                stack.append(loc_)
    ans += len(reachable)



print(f"part 1: {ans}")
print(f"part 2: {ans2}")
